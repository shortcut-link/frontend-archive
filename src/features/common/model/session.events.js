import { createEffect, createEvent } from 'effector';

export const sessionFetchProcessing = createEffect();

export const sessionChange = createEvent();
export const sessionRemove = createEvent();
