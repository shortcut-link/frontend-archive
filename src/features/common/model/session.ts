import { createEffect, createEvent, createStore, forward } from 'effector';

import { GetResponse, sessionAPI } from 'api/session';
import { User } from 'api/account';
import { tokenRemove } from './token';

export type options = { [key: string]: boolean };

export const sessionChange = createEvent<User>();
export const sessionRemove = createEvent<void>();
export const sessionFetchProcessing = createEffect<void, GetResponse>();
export const optionsChange = createEvent<options>();

export const $session = createStore<User | null>(null);

$session
  .on(sessionFetchProcessing.done, (_, { result: { user } }) => user)
  .on(sessionChange, (_, user) => user)
  .on(optionsChange, (user, options) => ({
    ...user,
    ...options
  }))
  .reset(sessionRemove);

sessionFetchProcessing.use(sessionAPI.get);

sessionFetchProcessing.fail.watch(() => sessionRemove());

forward({ from: sessionRemove, to: tokenRemove });
