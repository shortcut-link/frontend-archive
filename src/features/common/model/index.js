import './session.model';
import './token';
