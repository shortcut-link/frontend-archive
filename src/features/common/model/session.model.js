import { forward } from 'effector';

import { $session } from './session.store';
import {
  sessionFetchProcessing,
  sessionRemove,
  sessionChange
} from './session.events';
import { tokenRemove } from './token';
import { accountAPI } from 'api/account';

$session
  .on(sessionFetchProcessing.done, (_, { result: { user } }) => user)
  .on(sessionFetchProcessing.fail, () => null)
  .on(sessionChange, (_, user) => user)
  .reset(sessionRemove);

sessionFetchProcessing.use(() => accountAPI.getCurrentAccount());

forward({ from: sessionFetchProcessing.fail, to: tokenRemove });
forward({ from: sessionRemove, to: tokenRemove });
