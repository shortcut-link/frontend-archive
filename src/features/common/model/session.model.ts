import { forward } from 'effector';

import { $session } from './session.store';
import {
  sessionFetchProcessing,
  sessionRemove,
  sessionChange,
  optionsChange
} from './session.events';
import { tokenRemove } from './token';
import { sessionAPI } from 'api/session';

export type options = { [key: string]: boolean };

$session
  .on(sessionFetchProcessing.done, (_, { result: { user } }) => user)
  .on(sessionChange, (_, user) => user)
  .on(optionsChange, (user, options) => ({
    ...user,
    ...options
  }))
  .reset(sessionRemove);

sessionFetchProcessing.use(sessionAPI.get);

sessionFetchProcessing.fail.watch(() => sessionRemove());

forward({ from: sessionRemove, to: tokenRemove });
