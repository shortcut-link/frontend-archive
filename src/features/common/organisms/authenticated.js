//@flow
import { useStore } from 'effector-react';

import { $session, type Session as SessionType } from '../model/session.store';

type Props = {
  renderExists: (session: SessionType) => void,
  renderEmpty: () => void
};

export const Authenticated = ({ renderExists, renderEmpty }: Props) => {
  const session = useStore($session);
  if (session && renderExists) {
    return renderExists(session);
  } else {
    return renderEmpty();
  }
};
