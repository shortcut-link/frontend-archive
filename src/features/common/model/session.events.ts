import { createEffect, createEvent } from 'effector';
import { GetResponse } from 'api/session';
import { User } from 'api/account';
import { options } from './session.model';

export const sessionFetchProcessing = createEffect<void, GetResponse>();

export const sessionChange = createEvent<User>();
export const sessionRemove = createEvent<void>();

export const optionsChange = createEvent<options>();
