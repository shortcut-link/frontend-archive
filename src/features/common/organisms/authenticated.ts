import { useStore } from 'effector-react';

import { $session } from '../model/session';
import { User } from 'api/account';

interface AuthenticatedProps {
  renderExists: (session: User) => JSX.Element;
  renderEmpty: () => JSX.Element;
}

export const Authenticated = ({
  renderExists,
  renderEmpty
}: AuthenticatedProps): JSX.Element => {
  const session = useStore($session);

  if (session && renderExists) {
    return renderExists(session);
  } else {
    return renderEmpty();
  }
};
