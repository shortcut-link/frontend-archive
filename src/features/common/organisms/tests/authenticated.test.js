import React from 'react';
import { create, act } from 'react-test-renderer';

import { Authenticated } from '../authenticated';
import { sessionChange } from '../../model/session';

describe('must render Header', () => {
  const renderExists = () => <div>renderExists</div>;
  const renderEmpty = () => <div>renderEmpty</div>;

  const component = create(
    <Authenticated renderExists={renderExists} renderEmpty={renderEmpty} />
  );

  it('without session', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('with session', () => {
    act(() => {
      sessionChange({ email: 'example.com' });
    });

    expect(component.toJSON()).toMatchSnapshot();
  });
});
