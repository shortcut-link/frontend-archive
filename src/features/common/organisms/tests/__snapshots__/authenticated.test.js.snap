// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must render Header with session 1`] = `
<div>
  renderExists
</div>
`;

exports[`must render Header without session 1`] = `
<div>
  renderEmpty
</div>
`;
